src/outputs/restart.hpp

```diff
@@ -83,262 +64,62 @@ class RestartReader {
-      total_count *= shape[rank - 1 - i];
-    }
-    h5_count[rank] = count;
-    h5_offset[rank] = offset;
-    if (dataVec.size() < total_count) { // greedy re-alloc
-      dataVec.resize(total_count);
-    }
-    PARTHENON_HDF5_CHECK(H5Sselect_hyperslab(hdl.dataspace, H5S_SELECT_SET, h5_offset,
-                                             NULL, h5_count, NULL));
-    const H5S memspace = H5S::FromHIDCheck(H5Screate_simple(rank + 1, h5_count, NULL));
-    PARTHENON_HDF5_CHECK(H5Dread(hdl.dataset, hdl.type, memspace, hdl.dataspace,
-                                 H5P_DEFAULT, dataVec.data()));
-#endif // ENABLE_HDF5
-  }
-
-  // Reads an array dataset from file as a 1D vector.
-  template <typename T>
-  std::vector<T> ReadDataset(const std::string &name) const {
-#ifndef ENABLE_HDF5
-    PARTHENON_FAIL("Restart functionality is not available because HDF5 is disabled");
-#else  // HDF5 enabled
-    auto hdl = OpenDataset<T>(name);
-
-    std::vector<T> data(hdl.count);
-    const H5S memspace =
-        H5S::FromHIDCheck(H5Screate_simple(hdl.rank, hdl.dims.data(), NULL));
-
-    // Read data from file
-    PARTHENON_HDF5_CHECK(H5Dread(hdl.dataset, hdl.type, memspace, hdl.dataspace,
-                                 H5P_DEFAULT, static_cast<void *>(data.data())));
-
-    return data;
-#endif // ENABLE_HDF5
-  }
-
-  template <typename T>
-  std::vector<T> GetAttrVec(const std::string &location, const std::string &name) const {
-#ifndef ENABLE_HDF5
-    PARTHENON_FAIL("Restart functionality is not available because HDF5 is disabled");
-#else  // HDF5 enabled
-    // check if the location exists in the file
-    PARTHENON_HDF5_CHECK(H5Oexists_by_name(fh_, location.c_str(), H5P_DEFAULT));
-
-    // open the object specified by the location path, this could be a dataset or group
-    const H5O obj = H5O::FromHIDCheck(H5Oopen(fh_, location.c_str(), H5P_DEFAULT));
-
-    return HDF5ReadAttributeVec<T>(obj, name);
-#endif // ENABLE_HDF5
-  }
-
-  template <typename T>
-  T GetAttr(const std::string &location, const std::string &name) const {
-    // Note: We don't need a template specialization for std::string, since that case will
-    // be handled by HDF5ReadAttributeVec
-    auto res = GetAttrVec<T>(location, name);
-    if (res.size() != 1) {
-      PARTHENON_THROW("Expected a scalar attribute " + name +
-                      ", but got a vector of length " + std::to_string(res.size()));
-    }
-
-    return res[0];
-  }
+  virtual void ReadSwarmVar(const std::string &swarmname, const std::string &varname,
+                            const std::size_t count, const std::size_t offset,
+                            const Metadata &m, std::vector<Real> &dataVec) = 0;
+  virtual void ReadSwarmVar(const std::string &swarmname, const std::string &varname,
+                            const std::size_t count, const std::size_t offset,
+                            const Metadata &m, std::vector<int> &dataVec) = 0;
```

Comment:
Unfortunate that these can't remain templated... On the other hand, maybe this will speed up compile times...